### PR TITLE
fixes bug 1126776 - Analytics for the public API

### PR DESCRIPTION
**Pull request by peterbe**
CC @adngdb 

To test this manually, I ran this: https://gist.github.com/ccc7d2f3a4329febfd777f4441baa2f0
Then I edited my `.env` like this:
```
GOOGLE_ANALYTICS_API_URL=http://localhost:5000/?timeout=2
GOOGLE_ANALYTICS_DOMAIN=what.ever.org
GOOGLE_ANALYTICS_ID=ANY-THING
```

Things to note;

I'm still not 100% sure what to put for the `cid` value. In `fjord` the `cid` was usually made around a particular Django ORM object. That doesn't really exist in this context. The documentation on `cid` is both vague and clear at the same time. Generally, it's supposed to be like a session ID for the client. That doesn't exist when people do `curl https://crash-stats.m.c/api/SomeModel/` since "curlers" wouldn't re-use a cookie. We could do something like this:
```
cid = request.session.get('ga_cid')
if not cid:
    cid = uuid.uuid4().hex
    request.session['ga_cid'] = cid
params['cid'] = cid
```
But I'm not sure how I feel about modifying the `request` object in a tracking function like this. Also, our goal is to find out how many hits we get per Web API endpoint. Not who's surfing from one to another. 

~~I've tried to fire this against our stage domain, but nothing shows up when I'm logged in to Google Analytics Real-time Overview. Not sure if I'm expected to. The only way to know is to use it I suspect. Or wait a day and see if it appears in the aggregates.~~
Using my local curl and editing my `.env` to simulate stage's I really was able to trigger pageviews on API calls. See:
screen shot 2016-04-18 at 4 15 01 pm 
In other words! Unless google analytics later filter these out; it works!

I have tested this code when running my local django via uwsgi as a socket. Things worked just fine.

### CODE REVIEW

**Comment by adngdb** [comment]
:dancers:

File: webapp-django/crashstats/settings/test.py [hunk 1]
```diff
@@ -72,3 +72,9 @@
 
 MIDDLEWARE_RETRIES = 10
 MIDDLEWARE_MIDDLEWARE_RETRY_SLEEPTIME = 3
+
+# During testing, if mocking isn't done right, we never want to
+# accidentally send data to Google Analytics.
+GOOGLE_ANALYTICS_API_URL = 'https://fake.guugle-analytics.fake/collect'
```
Comment (willkg) [inline webapp-django/crashstats/settings/test.py]: It's always better to use "example.com" for fake domain names. That's what it's for.

File: webapp-django/crashstats/api/views.py [hunk 2]
```diff
@@ -318,6 +319,9 @@ def model_wrapper(request, model_name):
                 content_type='application/json; charset=UTF-8'
             )
 
+        # Execution was successful, inform Google Analytics
+        track_api_pageview(request)
```
Comment (willkg) [inline webapp-django/crashstats/api/views.py]: Why do this here as opposed to as a decorator on the view function?

File: webapp-django/crashstats/base/ga.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+# import sys
+# logger.addHandler(logging.StreamHandler(sys.stdout))
+
+
+def track_api_pageview(
+    request,
+    page_title=None,
+    data_source='api',
+    client_id=None,
+    async=True,
```
Comment (willkg) [inline webapp-django/crashstats/base/ga.py]: Why implement sync and async versions when you only want to use the async form?

File: webapp-django/crashstats/settings/test.py (hunk 1 above)
Comment (peterbe) [inline webapp-django/crashstats/settings/test.py]: fixed.

File: webapp-django/crashstats/api/views.py (hunk 2 above)
Comment (peterbe) [inline webapp-django/crashstats/api/views.py]: My original intent was to not trigger a tracking on failed things. But in afterthought I guess we can still use a decorator and trigger on the response if its status_code is 200.

Comment (willkg) [inline webapp-django/crashstats/api/views.py]: Is it interesting to know about failed uses of the api?

File: webapp-django/crashstats/settings/test.py (hunk 1 above)
Comment (willkg) [inline webapp-django/crashstats/settings/test.py]: This is a nit of mine because it's bit me several times.

In case anyone doesn't know and is reading along, there are a handful of top-level domains that are reserved for specific purposes like documenting and testing. Using these appropriately helps when reading the relevant code since the reader knows exactly what the thing is. Further, you can't accidentally have the code work with unintended consequences.

* http://www.iana.org/domains/reserved
* http://tools.ietf.org/html/rfc2606

Fun thing I just discovered: if you type "example.com" in Firefox, it goes to a nice page that talks about this.

**Comment by peterbe** [comment]
CC @adngdb @lonnen 
One interesting challenge is; do we want to track usage WITHIN the webapp too? (I think not) For example, the Home page uses the API in client-side but ultimately the user only loads `/home/product/Firefox` so it would be weird to say that also loading `/api/SuperSearch` is part of that pageview. 

I can accomplish this distinction by looking at the HTTP_REFERER which things like jQuery sets. 

If we don't do this, for every loading of the home page (with javascript enabled) would yield a count of 4 page views.

File: webapp-django/crashstats/base/ga.py (hunk 3 above)
Comment (peterbe) [inline webapp-django/crashstats/base/ga.py]: I thought I couldn't mock the async nature and since I know that tests would always need to be "eager" I thought this could come in handy. 
But I'll remove it now. It's silly to have code checked in until it's needed.

Comment (peterbe) [inline webapp-django/crashstats/base/ga.py]: removed now.

File: webapp-django/crashstats/crashstats/decorators.py [hunk 4]
```diff
@@ -65,3 +68,21 @@ def inner(request, *args, **kwargs):
             raise
         return view(request, *args, **kwargs)
     return inner
+
+
+def track_api_pageview_decorator(view):
```
Comment (adngdb) [inline webapp-django/crashstats/crashstats/decorators.py]: Isn't it weird to put ``decorator`` in the name of a decorator, in a ``decorators.py`` file?

Comment (peterbe) [inline webapp-django/crashstats/crashstats/decorators.py]: It is. I guess I could do `from crashstats.base.ga import track_api_pageview as _track_api_pageview` or something and call the decorator `track_api_pageview` instead. What do you think?

File: webapp-django/crashstats/base/ga.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        timeout=settings.GOOGLE_ANALYTICS_API_TIMEOUT
+    )
+
+    def success_cb():
+        # Note! This will trigger as long as there's not python exception
+        # happening inside the send.
+        # Meaning, if the requests.post(...).status_code != 200, this
+        # callback is still called.
+        logger.info(
+            'Successfully attempted to sent pageview to Google Analytics (%s)',
```
Comment (willkg) [inline webapp-django/crashstats/base/ga.py]: Nit: sent -> send.

File: webapp-django/crashstats/crashstats/decorators.py (hunk 4 above)
Comment (willkg) [inline webapp-django/crashstats/crashstats/decorators.py]: You could make `track_api_pageview` work as both a function and a decorator but that's probably more exotic than it's worth.

I think it's good to have two things that do different things have two different names. Beyond that, I don't think this matters much in the grand scheme of things.

File: webapp-django/crashstats/base/ga.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    params['dp'] = request.path  # Page
+    if request.META.get('QUERY_STRING'):
+        params['dp'] += '?{}'.format(request.META.get('QUERY_STRING'))
+
+    params['dt'] = page_title
+
+    if request.META.get('REMOTE_ADDR'):
+        ip = request.META['REMOTE_ADDR']
+        if ip != '127.0.0.1':
+            params['uip'] = ip
```
Comment (willkg) [inline webapp-django/crashstats/base/ga.py]: One thing this does is it subverts anyone's preferences for being tracked. If I don't want to be tracked, then I can prevent GA from loading in my browser and doing things. However, since we're doing this server-side, even if I do that, I'm still getting tracked.

Thus I'm not wildly excited about using peoples' ip addresses here.

With Input, I "faked" the id by using the request object's id. In that way, it's not tied to someone in a way that's identifying.

Do you have something similar you can use? A Django session id, perhaps?

Comment (willkg) [inline webapp-django/crashstats/base/ga.py]: Er, scratch that--Input doesn't provide a uid at all. It's optional. I suggest you not provide one either.

File: webapp-django/crashstats/base/ga.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    #   The value of this field should be a random UUID (version 4) as
+    #   described in http://www.ietf.org/rfc/rfc4122.txt
+    params['cid'] = client_id or uuid.uuid4().hex
+
+    if not request.user.is_anonymous():
+        params['uid'] = str(request.user.id)
+
+    params['dp'] = request.path  # Page
+    if request.META.get('QUERY_STRING'):
+        params['dp'] += '?{}'.format(request.META.get('QUERY_STRING'))
```
Comment (willkg) [inline webapp-django/crashstats/base/ga.py]: I think this is wrong. I think dp is just the path and shouldn't include the querystring portion. I think the querystring is garnered from dl by GA.

https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#dp

File: webapp-django/crashstats/base/ga.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        return
+
+    domain = settings.GOOGLE_ANALYTICS_DOMAIN
+    if not domain or domain == 'auto':
+        domain = RequestSite(request).domain
+
+    params = {}
+    params['v'] = 1  # version
+    params['tid'] = settings.GOOGLE_ANALYTICS_ID  # Tracking ID / Property ID
+    params['dh'] = domain
```
Comment (willkg) [inline webapp-django/crashstats/base/ga.py]: We might as well send dl since we're sending all the parts.

https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#dl

**Comment by willkg** [comment]
The changes look fine, so I spent some time working through what we're sending to GA in regards to PII and making sure we're tracking the information we intend to. I made comments accordingly.

I'm still a little puzzled regarding dl and dp. Seems like we shouldn't be sending query params with dp and instead send them with dl. If that's true, I'm not sure why it seems to work as is.

File: webapp-django/crashstats/crashstats/decorators.py (hunk 4 above)
Comment (peterbe) [inline webapp-django/crashstats/crashstats/decorators.py]: I think I'll go for `from crashstats.base.ga import track_api_pageview as _track_api_pageview` or something such.

File: webapp-django/crashstats/base/ga.py (hunk 6 above)
Comment (peterbe) [inline webapp-django/crashstats/base/ga.py]: Removed.

**Comment by peterbe** [comment]
@willkg Addressed. I removed the `uip` and added `dl` accordingly.

File: webapp-django/crashstats/base/ga.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    if not request.user.is_anonymous():
+        params['uid'] = str(request.user.id)
+
+    params['dp'] = request.path  # Page
+    params['dl'] = request.build_absolute_uri()
+
+    params['dt'] = page_title
+
+    if request.META.get('HTTP_USER_AGENT'):
+        params['ua'] = request.META['HTTP_USER_AGENT']
```
Comment (willkg) [inline webapp-django/crashstats/base/ga.py]: So, I know with Input, I would run PII-relevant things by @lonnen to get his :+1: because he was the all-powerful all-seeing data privacy sauron in the sky.

I don't know how that works with socorro--when we do PII-relevant things, does that require sign-off from someone who is one with the elemental nature of PII? If so, the above data we'd like to send to GA probably falls into that category of things.

**Comment by peterbe** [comment]
Oops. Deleted the wrong branch,.

File: webapp-django/crashstats/base/ga.py (hunk 9 above)
Comment (willkg) [inline webapp-django/crashstats/base/ga.py]: I talked with @lonnen and he said this is "probably fine". He can either chime in or I'll take the responsibility of assuming that means "completely fine" and we can move forward.

File: webapp-django/crashstats/base/ga.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        # happening inside the send.
+        # Meaning, if the requests.post(...).status_code != 200, this
+        # callback is still called.
+        logger.info(
+            'Successfully attempted to send pageview to Google Analytics (%s)',
+            params,
+        )
+
+    def failure_cb(exception):
+        logger.error('Failed to send GA page tracking (%s)', exception)
```
Comment (willkg) [inline webapp-django/crashstats/base/ga.py]: Is this actually an exception instance? Do you know offhand if the rest of `sys_excinfo` is available at this point or has it been corrupted by whatever else happened between the exception being thrown and this callback being called?

I ask because if everything is intact, you could do `logger.exception('Failed to send GA page tracking.')` and get better data including a traceback.

**Comment by willkg** [comment]
I had a question about the exception argument of `failure_cb`. I think everything else is covered. The `failure_cb` thing probably isn't a big deal, so if you want to land this today, it's fine as is.

What do you want to do?

**Comment by willkg** [comment]
To clarify, r+ with comments about the `failure_cb`.

**Comment by peterbe** [comment]
I need to play around with that `failure_cb` and see if `logger.exception` would work (because that assumes it gets something from `sys.exc_info()` right?)

**Comment by willkg** [comment]
During lunch, I thought about this some and I think we should just leave it as is.  Yes, we could look into this further to see if we could get traceback information, but honestly, I doubt we're ever going to look at the logs in this case and I'm pretty sure there's nothing that's going to trigger something that pokes us into doing so. So, it seems pretty moot and probably not worth spending time on right now. We can always go figure it out later.

Probably worth tossing in a FIXME comment, though.

**Comment by peterbe** [comment]
I checked. If you do something that triggers a `ConnectionError` for example (which I did by setting `GOOGLE_ANALYTICS_API_URL` to something that isn't running) it does call the `failure_cb` within the context of an exception. 

One final r?
